src/accelerate/utils/fsdp_utils.py

```diff
@@ -371,3 +424,149 @@ def param_init_fn_tied_param(module: torch.nn.Module):
+        )
+
+    fsdp2_kwargs = {
+        "reshard_after_forward": fsdp2_plugin.reshard_after_forward,
+        "offload_policy": fsdp2_plugin.cpu_offload,
+        # `fully_shard` doesn't accept `None` in case of `MixedPrecisionPolicy`
+        "mp_policy": fsdp2_plugin.mixed_precision_policy or MixedPrecisionPolicy(),
+    }
+
+    auto_wrap_policy = fsdp2_prepare_auto_wrap_policy(fsdp2_plugin, auto_wrap_policy_type, model)
+    if auto_wrap_policy is not None:
+        for module in get_children_bottom_up(model)[:-1]:  # We skip the model itself, as that one is always wrapped
+            if auto_wrap_policy(module):
+                fully_shard(module, **fsdp2_kwargs)
+
+    fully_shard(model, **fsdp2_kwargs)
+
+    if fsdp2_plugin.cpu_ram_efficient_loading:
+        # If `cpu_ram_efficient_loading` is enabled, only rank 0 loads the weights
+        # Other ranks have an empty model on `meta` device, so we need to distribute the weights properly
+        fsdp2_load_full_state_dict(accelerator, model, original_sd)
+
+    if accelerator.mixed_precision != "no" and model.dtype != torch.float32:
+        # We upcast the model according to `deepspeed`'s implementation
+        # More info about this can be found in `accelerator.py:prepare_model`s FSDP1 section
+        model = model.to(torch.float32)
+        if accelerator.is_main_process:
+            # TODO(siro1): Add a warning for each parameter that was upcasted
+            warnings.warn("FSDP upcast of low precision parameters may affect the precision of model checkpoints.")
+    return model
+
+
+def fsdp2_prepare_auto_wrap_policy(fsdp2_plugin, auto_wrap_policy_type: str, model: torch.nn.Module):
+    if auto_wrap_policy_type == "transformer":
+        no_split_modules = model._no_split_modules
+        if no_split_modules is None:
+            no_split_modules = []
+        transformer_cls_names_to_wrap = list(no_split_modules)
+        if fsdp2_plugin.transformer_cls_names_to_wrap is not None:
+            transformer_cls_names_to_wrap = fsdp2_plugin.transformer_cls_names_to_wrap
+        transformer_cls_to_wrap = set()
+
+        for layer_class in transformer_cls_names_to_wrap:
+            transformer_cls = get_module_class_from_name(model, layer_class)
+            if transformer_cls is None:
+                raise ValueError(f"Could not find the transformer layer class {layer_class} in the model.")
+            transformer_cls_to_wrap.add(transformer_cls)
+
+        def policy(module: torch.nn.Module) -> bool:
+            if fsdp2_plugin.transformer_cls_names_to_wrap is None:
+                return False
+            return isinstance(module, tuple(transformer_cls_to_wrap))
+
+    elif auto_wrap_policy_type == "size":
+
+        def policy(module: torch.nn.Module) -> bool:
+            return module.numel() > fsdp2_plugin.min_num_params
+    else:
+        return None
+
+    return policy
+
+
+def get_fsdp2_grad_scaler(**kwargs):
```

Comment:
(This one is fine, maybe just a comment explaining why we need to do this/where it's used)